cli/download_source.go

```diff
@@ -0,0 +1,259 @@
+// sha1 of the entire source URL without the query string. For remote URLs (e.g. Git URLs), this is based on the
+// assumption that the scheme/host/path of the URL (e.g. git::github.com/foo/bar//some-module) identifies
+// the module name and the query string (e.g. ?ref=v0.0.3) identifies the version. For local file paths, there is no
+// query string, so the same file path (/foo/bar) is always considered the same version. See also the encodeSourceVersion
+// and processTerraformSource methods.
+func encodeSourceName(sourceUrl *url.URL) (string, error) {
+	sourceUrlNoQuery, err := urlhelper.Parse(sourceUrl.String())
+	if err != nil {
+		return "", errors.WithStackTrace(err)
+	}
+
+	sourceUrlNoQuery.RawQuery = ""
+
+	return util.EncodeBase64Sha1(sourceUrlNoQuery.String()), nil
+}
+
+// Returns true if the given URL refers to a path on the local file system
+func isLocalSource(sourceUrl *url.URL) bool {
+	return sourceUrl.Scheme == "file"
+}
+
+// If this temp folder already exists, simply delete all the Terraform configurations (*.tf) within it
+// (the terraform init command will redownload the latest ones), but leave all the other files, such
+// as the .terraform folder with the downloaded modules and remote state settings.
+func cleanupTerraformFiles(path string, terragruntOptions *options.TerragruntOptions) error {
+	if !util.FileExists(path) {
+		return nil
+	}
+
+	terragruntOptions.Logger.Printf("Cleaning up existing *.tf files in %s", path)
+
+	files, err := filepath.Glob(filepath.Join(path, "*.tf"))
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	return util.DeleteFiles(files)
+}
+
+// There are two ways a user can tell Terragrunt that it needs to download Terraform configurations from a specific
+// URL: via a command-line option or via an entry in the .terragrunt config file. If the user used one of these, this
+// method returns the source URL and the boolean true; if not, this method returns an empty string and false.
+func getTerraformSourceUrl(terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig) (string, bool) {
+	if terragruntOptions.Source != "" {
+		return terragruntOptions.Source, true
+	} else if terragruntConfig.Terraform != nil && terragruntConfig.Terraform.Source != "" {
+		return terragruntConfig.Terraform.Source, true
+	} else {
+		return "", false
+	}
+}
+
+// Download the code from the Canonical Source URL into the Download Folder using the terraform init command
+func terraformInit(terraformSource *TerraformSource, terragruntOptions *options.TerragruntOptions) error {
+	terragruntOptions.Logger.Printf("Downloading Terraform configurations from %s into %s", terraformSource.CanonicalSourceURL, terraformSource.DownloadDir)
+
+	terragruntInitOptions := terragruntOptions.Clone(terragruntOptions.TerragruntConfigPath)
+	terragruntInitOptions.TerraformCliArgs = []string{"init", terraformSource.CanonicalSourceURL.String(), terraformSource.DownloadDir}
```

Comment:
Looks like `init` may be changed in a backwards incompatible way to be an interactive command. Not sure I can do much about that for now. One alternative is to try to use go-getter directly to reimplement the download functionality, but I'd rather wait and see if with the next version of Terraform, we can leverage `init` with special flags to do what we need.